desc/objectives/nae_utils.py

```diff
@@ -6,7 +6,79 @@
 from desc.grid import LinearGrid
 from desc.transform import Transform
 
-from .linear_objectives import FixSumModesR, FixSumModesZ
+from .linear_objectives import FixSumModesLambda, FixSumModesR, FixSumModesZ
+
+
+def calc_zeroth_order_lambda(qsc, desc_eq, N=None):
+    """Calculate 0th order NAE lambda constraint.
+
+    Uses the passed-in qsc object, and the desc_eq's stellarator symmetry is used.
+
+    Parameters
+    ----------
+    qsc : Qsc equilibrium
+        Qsc object to use as the NAE constraints on the DESC equilibrium.
+    desc_eq : Equilibrium
+        desc equilibrium to constrain.
+    N : int,
+        max toroidal resolution to constrain.
+        If None, defaults to equilibrium's toroidal resolution
+
+    Returns
+    -------
+    Lconstraints : tuple of Objective
+        tuple of constraints of type FixSumModesLambda, which enforce
+        the axis (O(rho^0)) behavior of the equilibrium lambda coefficents
+        to match the NAE.
+    L_0_n : array
+        array of floats of the toroidal Fourier coefficients of the
+        lambda on-axis behavior dictated by the NAE
+    Lbasis : FourierSeries Basis
+        FourierSeries basis corresponding to L_0_n coefficients
+
+    """
+    phi = qsc.phi
+    dphi = phi[1] - phi[0]
+    nfp = qsc.nfp
+    if N is None:
+        N = desc_eq.N
+    else:
+        N = np.min([desc_eq.N, N])
+    if desc_eq.sym:
+        Lbasis_sin = FourierSeries(N=desc_eq.N, NFP=nfp, sym="sin")
+    else:
+        Lbasis_sin = FourierSeries(N=desc_eq.N, NFP=nfp, sym=False)
+
+    grid = LinearGrid(M=0, L=0, zeta=phi, NFP=nfp)
+    Ltrans_sin = Transform(grid, Lbasis_sin, build_pinv=True, method="auto")
+
+    # from integrating eqn A20 in
+    # Constructing stellarators with quasisymmetry to high order 2019
+    #  Landreman and Sengupta
+    nu_0 = np.cumsum(qsc.B0 / qsc.G0 * qsc.d_l_d_phi - 1) * np.ones_like(phi) * dphi
+    nu_0_n = Ltrans_sin.fit(nu_0)
+
+    Lconstraints = ()
+    # lambda = -iota_0 * nu
+    L_0_n = -qsc.iota * nu_0_n
```

Comment:
are you sure this is correct? Eduardo and I spent a bunch of time going back and forth about the sign and don't remember where we landed. I think the "wrong" sign actually gave better results?

Reply:
This is the correct sign, we initialize it with the opposite sign when we do from_near_axis (I need to try changing that), but the final, correct lambda has this sign on it.